### PR TITLE
Expanding the supported schema of the Json Reader

### PR DESCRIPTION
**Pull request by ahsonkhan**
Adjusted the usability based on feedback.
Added more states to enable parsing of 'null', boolean, etc.

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/JsonReader.cs [hunk 1]
```diff
@@ -70,120 +89,218 @@ public bool Read()
...
+            MoveToNextTokenType();
+            return true;
+        }
+
+        public bool ReadTrue()
+        {
+            if (!Read())
+                throw new IndexOutOfRangeException("Json length is " + _str.Length + " and reading index " + _index +
+                                                   ".");
+            var utf8Bytes = new byte[4];
```
Comment (KrzysztofCwalina) [inline src/System.Text.Json/src/System/Text/Json/JsonReader.cs]: Why do you allocate the array?

File: src/System.Text.Json/src/System/Text/Json/JsonReader.cs [hunk 2]
```diff
@@ -70,120 +89,218 @@ public bool Read()
...
+            for (var i = 0; i < utf8Bytes.Length; i++)
             {
+                utf8Bytes[i] = (byte) _str[_index];
                 _index++;
             }
+            var utf8BytesString = new Utf8String(utf8Bytes);
```
Comment (KrzysztofCwalina) [inline src/System.Text.Json/src/System/Text/Json/JsonReader.cs]: why do you create this local? It's not used anywhere, is it?

File: src/System.Text.Json/src/System/Text/Json/JsonReader.cs (hunk 1 above)
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/JsonReader.cs]: This should be resolved after some of the new Utf8String APIs are in place, such as:
bool IsSubstringAt(index, Utf8String)

Comment (KrzysztofCwalina) [inline src/System.Text.Json/src/System/Text/Json/JsonReader.cs]: @krwq, don't you already have APIs to do this?

Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/JsonReader.cs]: Yes, you are right. The API I listed above is not yet available, but I can still use Substring(index, length) to achieve the same behavior without allocating.

Thanks.

File: src/System.Text.Json/tests/JsonReaderTests.cs [hunk 3]
```diff
@@ -8,153 +7,252 @@ namespace System.Text.Json.Tests
...
+                        };
+                        if (jsonPairs != null) jsonPairs.Add(pair);
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsString:
-                        value = json.ReadPropertyAsString();
+                    case JsonReader.JsonTokenType.Value:
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsInt:
-                        value = json.ReadPropertyValueAsInt();
+                    case JsonReader.JsonTokenType.Finish:
```
Comment (krwq) [inline src/System.Text.Json/tests/JsonReaderTests.cs]: Why not just return false from Read()?

File: src/System.Text.Json/tests/JsonReaderTests.cs [hunk 4]
```diff
@@ -8,153 +7,252 @@ namespace System.Text.Json.Tests
...
+        }
+
+        private static Array ReadArray(ref JsonReader jsonReader)
+        {
+            Array jsonArray = null;
+            List<Value> jsonValues = null;
+
+            while (jsonReader.Read())
+            {
+                switch (jsonReader.TokenType)
```
Comment (krwq) [inline src/System.Text.Json/tests/JsonReaderTests.cs]: I really think that Whitespace node should exist - it is easy filter them out but hard to add them when they are removed

File: src/System.Text.Json/tests/JsonReaderTests.cs [hunk 5]
```diff
@@ -8,153 +7,252 @@ namespace System.Text.Json.Tests
...
+                        break;
+                    case JsonReader.JsonTokenType.ArrayEnd:
+                        if (jsonArray != null)
+                        {
+                            var jsonElements = new Elements {Values = jsonValues};
+                            jsonArray.Elements = new List<Elements> {jsonElements};
+                            return jsonArray;
+                        }
+                        break;
+                    case JsonReader.JsonTokenType.Pair:
```
Comment (krwq) [inline src/System.Text.Json/tests/JsonReaderTests.cs]: Pair?

Comment (ahsonkhan) [inline src/System.Text.Json/tests/JsonReaderTests.cs]: A pair is a 'member' of an object and is of the form string:value.

http://json.org/
> An object is an unordered set of name/value pairs. An object begins with { (left brace) and ends with } (right brace). Each name is followed by : (colon) and the name/value pairs are separated by , (comma).

File: src/System.Text.Json/tests/JsonReaderTests.cs (hunk 4 above)
Comment (ahsonkhan) [inline src/System.Text.Json/tests/JsonReaderTests.cs]: Is this necessary? I am assuming the usage of Whitespace would be to maintain the input formatting in some way? What should the ReadWhitespace API return?

File: src/System.Text.Json/tests/JsonReaderTests.cs (hunk 3 above)
Comment (ahsonkhan) [inline src/System.Text.Json/tests/JsonReaderTests.cs]: You are right. I am already returning false from Read() and we never hit this case anyway.

**Comment by KrzysztofCwalina** [comment]
@ahsonkhan, it looks good to me. Let's merge and then do a quick API review (if you feel the APIs are reasonably ready)

**Comment by ahsonkhan** [comment]
I did not add have the Whitespace node. What are your thoughts on it? I don't think it is necessary to add it.

From @krwq 
> I really think that Whitespace node should exist - it is easy to filter them out but hard to add them when they are removed

From @ahsonkhan
> Is this necessary? I am assuming the usage of Whitespace would be to maintain the input formatting in some way? What should the ReadWhitespace API return?

**Pull request merged**